src/widgets/analytics/analytics.ts

```diff
@@ -241,6 +241,16 @@ const analytics: AnalyticsWidget = function analytics(widgetParams) {
         window.removeEventListener('beforeunload', onUnload);
       }
     },
+
+    getRenderState(renderState) {
+      return renderState;
+    },
+
+    getWidgetRenderState() {
+      return {
+        widgetParams,
+      };
+    },
```

Comment:
since getWidgetRenderState is optional, I think it could make sense for this widget not to have it, and leave it underpowered in favour for insights middleware

Reply:
Same for the `places` widget, no ?

Reply:
As long as we type those widgets like

```js
export type AnalyticsWidget = WidgetFactory<{}, {}, AnalyticsWidgetParams>;
```

TS will complain if there's no getWidgetRenderState() or getRenderState().

1. We can provide a dummy functions as-is
or
2. We can update their types to something like `() => Widget` (not providing `renderState` as a generic to Widget)

What do you think? @Haroenv @yannickcr

Reply:
Mmm, I'm not sure, but that's weird we have to do this for some optional methods :thinking: 

Question: How a user must type a custom widget ? And will he get the same issue ?

Reply:
that's because the WidgetRender params should be the last in the WidgetFactory type so you can type it as `none` here

Reply:
Not sure if I follow you @Haroenv 
If user is going to use `WidgetFactory` type instead of `Widget`, I don't see a way to not implement `getWidgetRenderState()` and `getRenderState`() without a TS error.

Reply:
I see what's going on now, since this is WidgetFactory, and not Widget, however if you put `never` for the `TRendererOptions` generic argument, it should mean you can't have the life cycles (as expected)

Reply:
putting `never` asked for 

```ts
getWidgetRenderState: (
  renderOptions
) => never;
```

I guess on the user land, they don't need to use `WidgetFactory`, which is not so beneficial to users:

```ts
export type WidgetFactory<TRendererOptions, TConnectorParams, TWidgetParams> = (
  widgetParams: TConnectorParams & TWidgetParams
) => Widget<{
  renderState: WidgetRenderState<TRendererOptions, TWidgetParams>;
}>;
```

I've updated the implementation just to use `Widget`.
What do you think?